[nethost] Adding a port for nethost

tannergooding opened the pull request:
This resolves https://github.com/microsoft/vcpkg/issues/8260.

tannergooding committed:
Adding a port for nethost

tannergooding commented:
This could likely do with some improvements, but it (as tested locally) should minimally support hosting netcoreapp3.0 compatible libraries.

tannergooding commented:
@PhoebeHui, what are the next steps required here?

PhoebeHui commented on ports/nethost/portfile.cmake:
```diff
@@ -0,0 +1,66 @@
+include(vcpkg_common_functions)
```
include(vcpkg_common_functions) doesn't need any more,  you can remove it.

PhoebeHui commented on ports/nethost/portfile.cmake:
```diff
@@ -0,0 +1,66 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO dotnet/core-setup
+    REF v3.0.0
```
Instead of 'REF v3.0.0', you can use 'REF 7d57652f33493fa022125b7f63aad0d70c52d810   #v3.0.0' since the commit is more stable than the revision.

PhoebeHui reviewed (commented):
@tannergooding, thanks for your contribution!

I noticed CI test failed on windows, the test log has gone, I will help rerun this PR after the CI test system is ready. Let's see which triplet failed here.

tannergooding commented:
I believe it was an arm job that failed with a note that it could be ignored if it wasnt supported (which it isnt currently).

PhoebeHui commented:
If it does not support arm build, you can add 'vcpkg_fail_port_install(ON_ARCH "arm" "arm64")' to portfile.cmake(Before vcpkg_from_github function), which the build should fail early.

tannergooding committed:
Fixing ref, removing unneeded include, and marking unsupported architectures for nethost

tannergooding commented:
Looks like there are now failures in unrelated ports.

vicroms commented:
Hi @tannergooding 

There are some issues with our CI system and we're working on fixing them, we'll rerun tests for your PRs once the CI system is back online.

JackBoosY commented:
/azp run

tannergooding commented:
Build failure is:
```
There are no build logs for nethost build.
This is usually because the build failed early and outside of a task that is logged.
See the console output logs from vcpkg for more information on the failure.
If the console output of the nethost is missing you can trigger a rebuild
in the test system by making any whitespace change to one of the files under
the ports/nethost directory or by asking a member of the vcpkg team to remove the
tombstone for abi tag 7734380cb0efc59353cf02ed1a771e7df46a12f4
```

Will retest locally tonight.

tannergooding commented:
Ah, nevermind. It's because I changed the `REF` from a tag to a commit and the hash needs updating.

tannergooding committed:
Fixing the expected sha512 for nethost

tannergooding committed:
Setting nethost to fail on static and uwp

tannergooding commented:
@PhoebeHui, I've marked the unsupported build scenarios (arm, arm64, uwp, and static) as unsupported; but it looks like this still fails CI. Is that expected?

JackBoosY commented on ports/nethost/portfile.cmake:
```diff
@@ -0,0 +1,66 @@
+vcpkg_fail_port_install(ON_ARCH "arm" "arm64" ON_LIBRARY_LINKAGE "static" ON_TARGET "uwp")
```
Please remove `ON_LIBRARY_LINKAGE "static"` and add `vcpkg_check_linkage(ONLY_STATIC_LIBRARY)`.

tannergooding replied:
That looks to do the opposite of what's intended?

This library is only meant to be compiled and consumed as shared, so I believe `vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)` would be the appropriate condition (looking at the docs).

tannergooding replied:
I've pushed `ONLY_DYNAMIC_LIBRARY` for now, but will happily correct if I was mistaken 😄

JackBoosY replied:
Sorry, my fault. Your changes is good.

JackBoosY commented:
In x86-windows:
```
FAILED: ../CMakeCache.txt 
cmd /c "cd .. && "D:/downloads/tools/cmake-3.14.0-windows/cmake-3.14.0-win32-x86/bin/cmake.exe" "C:/vsts/_work/2/s/buildtrees/nethost/src/d70c52d810-d1c9f88a20/src/corehost/cli/nethost" "-DSKIP_VERSIONING=1" "-DCLI_CMAKE_HOST_POLICY_VER:STRING=3.0.0" "-DCLI_CMAKE_HOST_FXR_VER:STRING=3.0.0" "-DCLI_CMAKE_HOST_VER:STRING=3.0.0" "-DCLI_CMAKE_COMMON_HOST_VER:STRING=3.0.0" "-DCLI_CMAKE_PKG_RID:STRING=win-x86" "-DCLI_CMAKE_COMMIT_HASH:STRING=vcpkg release/3.0.0" "-DCLI_CMAKE_PLATFORM_ARCH_I386=1" "-DCMAKE_MAKE_PROGRAM=C:/vsts/_work/2/s/downloads/tools/ninja/ninja-1.8.2/ninja.exe" "-DBUILD_SHARED_LIBS=ON" "-DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=C:/vsts/_work/2/s/scripts/toolchains/windows.cmake" "-DVCPKG_TARGET_TRIPLET=x86-windows" "-DVCPKG_SET_CHARSET_FLAG=ON" "-DVCPKG_PLATFORM_TOOLSET=v142" "-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON" "-DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE" "-DCMAKE_VERBOSE_MAKEFILE=ON" "-DVCPKG_APPLOCAL_DEPS=OFF" "-DCMAKE_TOOLCHAIN_FILE=C:/vsts/_work/2/s/scripts/buildsystems/vcpkg.cmake" "-DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON" "-DVCPKG_CXX_FLAGS=" "-DVCPKG_CXX_FLAGS_RELEASE=" "-DVCPKG_CXX_FLAGS_DEBUG=" "-DVCPKG_C_FLAGS=" "-DVCPKG_C_FLAGS_RELEASE=" "-DVCPKG_C_FLAGS_DEBUG=" "-DVCPKG_CRT_LINKAGE=dynamic" "-DVCPKG_LINKER_FLAGS=" "-DVCPKG_TARGET_ARCHITECTURE=x86" "-DCMAKE_INSTALL_LIBDIR:STRING=lib" "-DCMAKE_INSTALL_BINDIR:STRING=bin" "-G" "Ninja" "-DCMAKE_BUILD_TYPE=Release" "-DCMAKE_INSTALL_PREFIX=C:/vsts/_work/2/s/packages/nethost_x86-windows""
CMake Error: The source directory "C:/vsts/_work/2/s/buildtrees/nethost/src/d70c52d810-d1c9f88a20/src/corehost/cli/nethost" does not exist.
Specify --help for usage, or press the help button on the CMake GUI.
ninja: build stopped: subcommand failed.
```
Why?

In x64-windows:
```
-- Note: nethost only supports dynamic library linkage. Building dynamic library.
CMake Error at scripts/cmake/vcpkg_check_linkage.cmake:42 (message):
  Refusing to build unexpected dynamic library against the static CRT.  If
  this is desired, please configure your triplet to directly request this
  configuration.
```

I think we should revert changes about `ONLY_DYNAMIC_LIBRARY`, use `vcpkg_fail_port_install(ON_LIBRARY_LINKAGE "static")` is expected.

PhoebeHui commented:
@tannergooding, arm, arm64, uwp, and static CI failed as expected, however, we need figure out why it failed on x86-windows, the failures looks from 'CMake Error: The source directory "C:/vsts/_work/2/s/buildtrees/nethost/src/d70c52d810-d1c9f88a20/src/corehost/cli/nethost" does not exist.' could you have a look please?

tannergooding commented:
Reverted the change using `vcpkg_check_linkage` and am back to using `vcpkg_fail_port_install(ON_LIBRARY_LINKAGE "static")` as was recommended by @JackBoosY.

JackBoosY commented on ports/nethost/usage:
```diff
@@ -0,0 +1,4 @@
+The package nethost provides CMake integration:
+
+    find_package(nethost CONFIG REQUIRED)
+    target_include_directories(main PRIVATE ${NETHOST_INCLUDE_DIRS})
```
Please add an empty line end of this file.

tannergooding replied:
This looks like an issue on the GitHub rendering side. The file (all of the ones I added) already have a newline at the end of the file.

If you click the three dots and choose edit file, you can see that the file has a line ending. You can also see this by manually checking out the relevant branch/commit.

JackBoosY commented on ports/nethost/portfile.cmake:
```diff
@@ -0,0 +1,66 @@
+        "-DCLI_CMAKE_PLATFORM_ARCH_${ARCH_NAME}=1"
+)
+
+file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/bin ${CURRENT_PACKAGES_DIR}/debug ${CURRENT_PACKAGES_DIR}/include ${CURRENT_PACKAGES_DIR}/lib)
```
Why delete the directory `bin` / `include` /  `lib` here?

tannergooding replied:
I believe there was some error that was occurring without this. IIRC, it was related to trying a dynamic build and then a static build; but static builds are blocked now, so this might be unnecessary now.

tannergooding replied:
I've removed

tannergooding committed:
Removing an unecessary file(REMOVE_RECURSE) from nethost

tannergooding commented:
Hmmm, one of the x64 Windows legs failed because it couldn't find the relevant source directory.

However, its indicating that patches were successfully applied and there are no extraction errors. Other steps (like install) also completed sucesfully, just not the config step.

JackBoosY commented:
/azp run

JackBoosY commented:
/azp run

tannergooding commented:
Rebased onto current head. That should hopefully kick off the jobs again.

Commit:
update baseline

tannergooding commented:
@JackBoosY, what's still need for this to get merged now?

Also, thanks for the help in getting this resolved, etc 😄

dan-shaw commented:
LGTM, right now our MacOS CI pipeline is down, so we are waiting for that before merging this.

ras0219-msft committed:
[nethost] Update to latest from dotnet/runtime

ras0219-msft committed:
[vcpkg-baseline] Update baseline for nethost

ras0219-msft reviewed (approved):
Thanks for the PR! While reviewing, I made a few changes:

1. The dotnet/core-setup repo appears to be deprected in favor of dotnet/runtime, so I've switched sources
2. The general policy of vcpkg is to follow the latest stable release; however, I don't know what compatibility policies dotnet has w.r.t. this nethost component. I've updated to latest under the assumption that nethost is backwards compatible. If this doesn't work for your use cases, we'll need to discuss more about what's appropriate here (and how to handle other users which will want to use different dotnet versions).
3. A benefit of updating to the latest is that static builds are now supported upstream!
4. Another benefit is support for arm and arm64 :)
5. This package doesn't install any CMake integration files, so I've changed the usage instructions to use lower level mechanisms.

ras0219-msft committed:
[vcpkg-baseline] Update baseline for nethost

tannergooding commented:
> The dotnet/core-setup repo appears to be deprected in favor of dotnet/runtime, so I've switched sources

It's not that its deprecated. In fact, it will continue to be used for servicing .NET Core 3.1 and prior versions. dotnet/runtime is going to be used, moving forward, for .NET 5 and future versions (and is a consolidation of dotnet/corefx, dotnet/runtime, and dotnet/core-setup, primarily)

> The general policy of vcpkg is to follow the latest stable release;

The latest stable release is https://github.com/dotnet/core-setup/tree/v3.1.0. We won't be shipping a new stable release out of dotnet/runtime until next November (approx) when we plan on shipping .NET 5. -- You can see the public roadmap we've published here: https://github.com/dotnet/core/blob/master/roadmap.md

> I've updated to latest under the assumption that nethost is backwards compatible.

We should be generally backwards compatible provided the dev has supplied a `runtimeconfig.json` which specifies the desired target version and has that runtime installed on the box.

> A benefit of updating to the latest is that static builds are now supported upstream!
> Another benefit is support for arm and arm64 :)

I don't believe we've made any significant changes here between when we shipped 3.1 at the beginning of the month and now, so I might just have not configured this correctly when it was using `dotnet/core-setup`.

ras0219-msft commented:
/azp run

tannergooding commented:
Is there an ETA on when the OSX runs will be back online?

tannergooding commented:
Looks like CI finally ran but it failed when downloading cmake.

dan-shaw commented:
LGTM, we are going to merge some PRs without waiting for the MacOS CI to be back up
